packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,285 @@
+          staticData: {},
+        })
+
+      const key = pageData.key ?? 'main'
+      if (pageData.dataPath) {
+        if (page.data[key])
+          throw Error(
+            `setPageData conflict: data with key "${key}" already exists`
+          )
+        page.data[key] = pageData.dataPath
+      }
+      if (pageData.staticData) {
+        if (page.staticData[key])
+          throw Error(
+            `setPageData conflict: staticData with key "${key}" already exists`
+          )
+        page.staticData[key] = pageData.staticData
+      }
+
+      if (existingPage) {
+        changedPages.add(pageId)
+        emitChange()
+      }
+    }
+
+    const removePageData = (pageId: string, key?: string) => {
+      const page = pageCache[pageId]
+      if (page) {
+        if (key) {
+          delete page.data[key]
+          delete page.staticData[key]
+          changedPages.add(pageId)
+          emitChange()
+        } else {
+          delete pageCache[pageId]
+        }
+      }
+    }
+
+    const watchFiles = (
+      baseDir: string,
+      arg2?: string | string[] | FileHandler,
+      arg3?: FileHandler
+    ) => {
+      let globs: string[]
+      let handler: (
+        file: File,
+        type: string
+      ) => PageData | Promise<PageData> | void
+
+      if (typeof arg2 === 'function') {
+        globs = ['**/*']
+        handler = arg2
+      } else {
+        globs = Array.isArray(arg2) ? arg2 : [arg2 || '**/*']
+        handler = arg3 || defaultPageLoader
+      }
+
+      // Strip trailing slash and make absolute
+      baseDir = path.resolve(pagesDir, baseDir)
+
+      const processFile = async (
+        type: 'add' | 'change' | 'unlink',
+        filePath: string
+      ) => {
+        filePath = path.join(baseDir, filePath)
+
+        const file =
+          fileCache[filePath] ||
+          (fileCache[filePath] = new File(filePath, baseDir))
+
+        file.pageIds.forEach((pageId) => {
+          delete pageCache[pageId]
+        })
+
+        if (type === 'change') {
+          file.content = null
+          file.pageIds.clear()
+        }
+
+        const result =
+          handler.length > 1 || type !== 'unlink'
+            ? await handler(file, type)
+            : null
+
+        if (type === 'unlink') {
+          delete fileCache[file.path]
+        } else if (result) {
+          file.pageIds.add(result.pageId)
+          setPageData(result.pageId, result)
+        }
+      }
+
+      watchers.add(
+        chokidar
+          .watch(globs, {
+            cwd: baseDir,
+            ignored: ['**/node_modules/**/*'],
+          })
+          .on(
+            'all',
+            (type, filePath) =>
+              type !== 'addDir' &&
+              type !== 'unlinkDir' &&
+              onPageUpdate(processFile(type, filePath))
+          )
+      )
+    }
+
+    const helpers: PageHelpers = {
+      extractStaticData,
+      loadPageData: (file) => Promise.resolve(loadPageData(file, helpers)),
```

Comment:
It is a little wired that we get `loadPageData` from user, and the only place we use it is "provide it back to user"!